### PR TITLE
Optimise ExplicitSize performance

### PR DESCRIPTION
**Pull request by kulyk**
The new navigation sidebar has made the dashboard rendering performance problem more prominent. Opening/closing the navbar on a busy dashboard a few times in a row could end up with a very poor UX. The root cause is how we handle content dimension changes with the `ExplicitSize` HOC. It wasn't that noticeable until we had non-changeable dashboard width (ie the only way to resize the dashboard was to resize the browser window), but after we've added the new navigation bar, the problem is much more crucial.

### Problem

**Number of listeners**

The dashboard grid component is wrapped with `ExplicitSize` and the same goes for every visualization. Moreover, a single dashboard card can have a few components wrapped with `ExplicitSize` (e.g. `Visualization` + `CardVisualization` + `LegendHeader`).

As a result, if I x-ray the sample database's "Products" table, the automatic dashboard will have 55 `ExplicitSize` instances (and it's just 13 cards).

**Handling resizes**

`ExplicitSize's` main purpose is to measure the underlying component and pass `width` and `height` props to it. It uses [`ResizeObserver`](https://developer.mozilla.org/en-US/docs/Web/API/ResizeObserver) [polyfill](https://github.com/que-etc/resize-observer-polyfill) based on [`MutationObserver`](https://developer.mozilla.org/en-US/docs/Web/API/MutationObserver) to monitor element resizes. When the underlying component gets resized, `ExplicitSize` will remeasure it, update `width` and `height` in state and pass them down to the underlying component. The resize listener is throttled with a 300ms delay, so things don't get crazy. However, it will remeasure the component on every `componentDidUpdate` invocation.

### Solution

1. Stopped re-measuring dimensions on every `componentDidUpdate` invocation
2. Switched to real `ResizeObserver` API (it's available in 90% of browsers, compared to 95% of `MutationObserver`)
3. Added a shared `ResizeObserver` with `subscribe` / `unsubscribe` API, so all the `ExplicitSize` instances use a shared `ResizeObserver` instance. See details: https://github.com/WICG/resize-observer/issues/59
4. Wrapped the main `ResizeObserver` handler with [`raf-schd`](https://github.com/alexreardon/raf-schd) — a throttle utility based on `requestAnimationFrame`. See [why](https://github.com/alexreardon/raf-schd#why)
5. Extended `ExplicitSize`, so a component can configure if it needs to throttle or debounce the resize handler. This allows us to keep some necessary behaviors + optimize the others:
  * it's important to throttle `DashoardGrid's` listener, so you can get visual feedback when resizing a browser window. If we debounce it, a user will only see the end layout after they stop resizing
  * it's better to debounce dashboard card contents listeners though. In that way, you'll still see how the dashboard will look while you're resizing a window, but we'll be able to skip a few dimension updates and free up the main thread
  * for `TableSimple` we ideally want a debounce when resizing a dashboard and throttle when resizing the table card when editing a dashboard. E.g. #21743 made `ExplicitSize` always use debounce and it led to unexpected resizing behavior for tables, see [this comment](https://github.com/metabase/metabase/pull/21743#pullrequestreview-952231571)

### Visual Demo

⚠️ For demo purposes, I'm throttling the CPU by 4x, so the difference is more noticeable.

<details>
<summary>Video</summary>

**Before**

https://user-images.githubusercontent.com/17258145/165534026-06afca9c-0ae5-4003-aa43-3007ed82c9b0.mp4

**After**

https://user-images.githubusercontent.com/17258145/165534065-3266305d-6d06-4fee-b270-f2c7f6cfc93f.mp4

</details>

<details>
<summary>Performance Report</summary>

👀  The slowest task before was taking ~2.8 **seconds**. After the fix, the slowest task is around 815ms. We still have some room for improvement, but I'm not sure if there's a potential easy win in performance here now.

**Before**
before

**After**
after

</details>

### CODE REVIEW

File: frontend/test/__support__/mocks.js [hunk 1]
```diff
@@ -8,6 +8,12 @@ global.window.matchMedia = () => ({
   removeListener: () => {},
 });
 
+global.window.ResizeObserver = class ResizeObserver {
```
Comment (ranquild) [inline frontend/test/__support__/mocks.js:11]: Didn't get it why we need a class here and not an object. Could you explain?

**Review by ranquild — APPROVED** [review]
Great work! Left some comments

File: frontend/src/metabase/components/ExplicitSize.jsx [hunk 2]
```diff
@@ -47,7 +62,12 @@ export default ({ selector, wrapped } = {}) => ComposedComponent =>
     componentDidUpdate() {
       // update ResizeObserver if element changes
       this._updateResizeObserver();
-      this._updateSize();
+      if (typeof refreshMode === "function" && !isCypressActive) {
```
[the remark this answers is not in the document]
Comment (kulyk) [reply frontend/src/metabase/components/ExplicitSize.jsx:65]: Oh right, definitely missed that! Thanks

[the remark this answers is not in the document]
Comment (kulyk) [reply frontend/src/metabase/components/ExplicitSize.jsx:65]: Fixed ✨

**Review by daltojohnso — APPROVED** [review]
🥇

**Pull request merged**